### ISSUE TITLE
Liquid to Array

### ISSUE THREAD

**Issue by Warloxx** [cantino/huginn#1647]
I just made an Agent chain that results in one event with an array of urls as payload.
```javascript
{
  "urls": [
    "https://www.youtube.com/feeds/videos.xml?channel_id=UC5IMpS_PUb-3YDdLoH4XprA",
    "https://www.youtube.com/feeds/videos.xml?channel_id=UCtwKon9qMt5YLVgQt1tvJKg",
   ...,
   ...
  ]
}
```
This is fed into an Commander Agent that should update the "url" property of an Rss Agent which then pulls all the feeds.

BUT I can't get the urls array in the configure_options datastructure of the Commander Agent to update the config because liquid arguments only exists as strings.

I really would like to do something like this in the Commander Agent:
```javascript
{
  "action": "configure",
  "configure_options": {
    "expected_update_period_in_days": "5",
    "clean": "false",
    "url": {{urls}}
  }
}
```
_Notice_: no quotes around the {{urls}}.

I can't think of a way to change the url array of the Rss Agent with an arbitrary number of feeds. AFAIK the javascript Agent can't manipulate other agents, since that would be a somewhat easy workaround.

I found #1127 that contains similarities but seems to ultimately come down to the fact that the intermediary steps involved arrays but the end result was a string again. 

Any ideas?

**Comment by cantino** [comment]
Interesting, that does seem like something you should be able to do, but may not be possible right now.

As a side note, it would also be reasonable for us to change the RSSAgent so that it can fetch based on an incoming event, like the PostAgent and WebsiteAgent do.  For now, could you send your event to a WebsiteAgent and use it to pull the RSS?

**Comment by Warloxx** [comment]
I can't really feed the array into a Website Agent either, its the same problem again since it expects an array of strings with urls. I could only put in a string with the liquid formating. What I could do is to interface it with a javascript agent that converts the array into multiple events with a single string. This should work but it not as nice as it could be.

~~Now another problem that makes this unfeasible:
The website agents extractors need to have the same amount of items to output good events. But when scraping RSS that is not always true. At least in my use case I need the title of the RSS feed (/feed/title) to output in every event but it will only be matched once by the website agent since it is not included in every rss entry.~~

I just saw that the title of the RSS feed is, at least my case (youtube sub feeds), included as /entry/author/name.
So I think I can work around this. I'll report back whether or not it works.

**Comment by Warloxx** [comment]
I got it "working" now. I've created a JavaScript Agent that pulls the array of urls apart in multiple events with each a single url. That flows in the WebsiteAgent which scraps all the data. 

After this I created another JS Agent to filter all the entries from the feeds. Since I can't use the RSS reader I have to also remember which once have already been send out since the WebsiteAgent just recreates the events every time it runs.

In theory I have it working, BUT I have a problem with storing the data in memory. I have an object looks like this:
```
channels = {
   'channel1_name': [
      { 'id': '<id1>', 'date': '<date-sting>'},
      { 'id': '<id2>', 'date': '<date-sting>'}
   ],
   'channel2_name': [
      { 'id': '<id1>', 'date': '<date-sting>'},
      { 'id': '<id2>', 'date': '<date-sting>'}
   ]
}
```

but when I save this object in memory `this.memory('channels', channels);` it apparently saves it as a string. In the details view it looks like this:
```
{
  "channels": "[object Object]"
}
```

I checked the type with `typeof`, I definitely save an object yet it saves it as a string. And when I recall it I just get the string, which has no usable data.

I found [this](https://github.com/cantino/huginn/blob/master/spec/models/agents/java_script_agent_spec.rb#L173), and if I read line 173 right it should be possible to store objects.

**Comment by cantino** [comment]
That's weird.  I agree that the spec you linked to would make it seem like an object would work fine.  Could you share the JavaScript?

**Comment by Warloxx** [comment]
Here is the JS Code. I hope this makes sense. JS is not really a language I have used much before. I have left some comments in that show what else I have tested, like storing the object as an JSON serialized string. Storing worked fine but loading resulted in an error of `JavaScript error: Validation failed: Message can't be blank`. Googling that gave me a lot of results from ruby rather then JS, so I'm not sure if the error came from my JS code or the JS Agent executing it.

```Javascript
var channelsAllVideos = [
   "channel name1",
   "channel name2",
   ....SNIP....
];
Agent.IsNew = function(channels, videoData)
{
  if(!channels[videoData.channel_name])
    channels[videoData.channel_name] = [];
  
  var foundIndex = -1;
  // find video
  for(var i = 0; i< channels[videoData.channel_name].length; i++)
  {
    if(channels[videoData.channel_name][i].id == videoData.video_id)
    {
      foundIndex = i;
      break;
    }
  }
  // .findIndex() not supported
  /*foundIndex = channels[videoData.channel_name].findIndex(
    function(d)
    {
      return d.id == videoData.video_id;
    }
  );*/
  
  // new? save id and date
  if(foundIndex === -1)
  {
    channels[videoData.channel_name].push({'id': videoData.video_id, 'date': videoData.published });
  }
  
  //cleanup. only store the newest 15 video IDs (more are not provided with RSS)
  //sort by date:
  channels[videoData.channel_name].sort(
    function(a, b) {
      return new Date(a.date).getTime() - new Date(b.date).getTime();
    }
  );
  //oldest are at the bottom of the list now
  //Only keep max 15 entries
  while(channels[videoData.channel_name].length > 15)
  {
    channels[videoData.channel_name].pop();
  }
  
  return foundIndex === -1;
}

Agent.receive = function() {
  // get memory
  var channels = this.memory('channels');
  this.log(typeof channels);
  /*
  this.log(channels);
  if(channels != null)
    channels = JSON.parse(channels);
  else
    channels = {};
  */
    
  if(!channels)
    channels = {};
    
  this.log(typeof channels);
  
  var events = this.incomingEvents();
  for(var i = 0; i < events.length; i++) {
    var obj = events[i].payload;
    
    if(this.IsNew(channels, obj))
    {
      // enqueue everything from these channels
      if(channelsAllVideos.indexOf(obj.channel_name))
      {
        this.createEvent({ 'message': obj.video_id });
      }
      // more specific filter
      else if(obj.channel_name == "<some other channel name>" && !obj.video_title.toLowerCase().includes("<interesting title>"))
      {
        this.createEvent({ 'message': obj.video_id });
      }
      //SNIP; some more filter
    }

  }
  
  //save all the data back
  //this.memory('channels', JSON.stringify(channels));
  this.log(typeof channels);
  this.memory('channels', channels);
}
```

For context, the incoming payload of an event will have this form:
```Javascript
{
  "video_id": "...",
  "video_url": "...",
  "video_title": "...",
  "published": "...",
  "channel_name": "...",
  "channel_url": "..."
}
```

**Comment by dsander** [comment]
@Warloxx The code looks good to me, when did you last update Huginn? The change that fixed putting javascript objects in the memory is from the beginning if July.

**Comment by Warloxx** [comment]
Turns out I was running a version from 30.4.2016  #(1411)
And this doesn't have the capability to store arrays and objects.

Updating now, reporting back later.

**Comment by Warloxx** [comment]
After the update I can store my object just as expected in the memory. My workaround for the original problem works fine now.

Should this issue be closed now or kept open to keep track of the original issue not being able to pass on arrays via liquid formatting?

**Comment by dsander** [comment]
It would be great to have that functionality, but don't know if it is possible since liquid is a templating language and thus always returns strings.

**Comment by cantino** [comment]
This is the same as #1127, I believe.  It would be nice if we could convert Liquid returning an array into an actual array instead of a joined string.

**Comment by Warloxx** [comment]
Yea it looks like it, although, just like here, it got a bit off topic. 

If the liquid engine would be executed on the entire config string then this could work right? Although I imagine that would be a mayor change.

**Comment by dsander** [comment]
@Warloxx When you update Huginn to the latest version you should be able to use the new `as_object` filter in your CommanderAgent:

```
{
  "action": "configure",
  "configure_options": {
    "expected_update_period_in_days": "5",
    "clean": "false",
    "url": "{{urls | as_object}}"
  }
}
```

Feel free to reopen the issue if that is not working as intended.

**Comment by Warloxx** [comment]
just updated and tested it. Works as intended. Very nice.

Nasty side effect, I just hit the 500 IDs limit of the RSS agent :(
Fixed it: #1733
